### ISSUE TITLE
fitting discrete distributions

### ISSUE THREAD

**Issue by manuelsmendoza** [scipy/scipy#11948]
Actually we can use `scipy.stats.rv_continuous.fit` method to extract the parameters for a theoretical continuous distribution from empirical data, however, it is not implemented for discrete distributions e.g. negative binomial and Poisson... may it be implemented in a near future?

**Comment by jmmanley** [comment]
I agree this would be quite nice to have for the discrete distributions in `scipy.stats.rv_discrete`, and I would be happy to work on a PR if people think it's worthwhile. For a number of the current distributions (e.g. Poisson obviously), there are closed form solutions to the maximum likelihood estimates, others need optimizers / iterative algorithms, e.g. beta-binomial.

**Comment by mdhaber** [comment]
@WarrenWeckesser thought you should be included here.

I'd be happy to help review the discrete `fit` method PRs. Note that there is currently a lot of work in progress on the continuous side of things. gh-11695 adds method of moments, and @swallan has been looking into which distributions have closed form solutions for MLE and MM fits in gh-11782.  Would someone be willing to help review gh-11695?

**Comment by rlucas7** [comment]
> On Apr 27, 2020, at 8:36 PM, Matt Haberland <notifications@github.com> wrote:
> 
> ﻿
> @WarrenWeckesser thought you should be included here.
> 
> I'd be happy to help review the discrete fit method PRs.
> 

Note some distributions have some fairly complicated procedures and constraints. For example, a binomial distribution with a known success probability but unknown n parameter is non standard and I’m not sure how to handle the case if both p and n are not specified, there is probably a paper somewhere on that case though. 

I’m not opposed to the idea but it would be nice to see some research h into which discrete distributions we have that have closed form estimators, which require a root finding, and which are (currently) unknown.
IIRC there are currently 14 or so discrete distributions so shouldn’t be too excessive to research. 

If we agree to support we’d want to make sure we could support the list we have for discrete. 

 If others agree that support for discrete distribution fits should be supported then I’d recommend open a tracking issue with distributions listed and pursue one distribution per PR, e.g. one distribution at a time. 
> Note that there is currently a lot of work in progress on the continuous side of things. gh-11695 adds method of moments, and @swallan has been looking into which distributions have closed form solutions for MLE and MM fits in gh-11782.
> 
@mdhaber

>Would someone be willing to help review gh-11695?

If I have time this weekend on Sunday I’ll take another look at 11695.

**Comment by rlucas7** [comment]
discrete distributions taken from here:
https://github.com/scipy/scipy/blob/master/scipy/stats/_distr_params.py#L115

| dist name | mle | mme | 
| ----------|-------|------|
| bernoulli | [sample mean](https://en.wikipedia.org/wiki/Bernoulli_distribution#Properties) |. |
|betabinom| |[mme](https://en.wikipedia.org/wiki/Beta-binomial_distribution#Point_estimates) |
|binom| | [n + p](https://www.stat.purdue.edu/~dasgupta/publications/binomialn.pdf) [only p](https://en.wikipedia.org/wiki/Binomial_distribution#Estimation_of_parameters) [only n*](https://scholar.google.com/scholar?client=firefox-b-1-d&um=1&ie=UTF-8&lr&q=related:zzMHN3mV2lpqyM:scholar.google.com/) |
|boltzmann| see note |. |
|dlaplace|[mle^2](https://www.sciencedirect.com/science/article/abs/pii/S0378375804003519) |. |
|geom|[mle](https://en.wikipedia.org/wiki/Geometric_distribution#Parameter_estimation) |. |
|hypergeom|. |. |
|logser| |. |
|nbinom|[mle^1](https://en.wikipedia.org/wiki/Negative_binomial_distribution#Statistical_Inference) |. |
|planck| |. |
|poisson|[mle](https://en.wikipedia.org/wiki/Poisson_distribution#Parameter_estimation) | |
|randint|[mle for upper bound](https://en.wikipedia.org/wiki/Discrete_uniform_distribution#Estimation_of_maximum) |. |
|skellam|[mle^3](https://faculty.georgetown.edu/kfs7/MY%20PUBLICATIONS/Sellers%202012%20AASS001.pdf) |. |
|zipf| |. |
|yulesimon| [mle](https://en.wikipedia.org/wiki/Yule%E2%80%93Simon_distribution) | [only if >1](https://en.wikipedia.org/wiki/Yule%E2%80%93Simon_distribution) | 

* note there are several proposals for the 'only n' case. 
^1 note the UMVUE dominates the MLE, and MLE is biased, and ***is not*** the MME 
^2 MLE `\hat{p}= \frac{T}{1+\sqrt{1+T}}` where `T=\sum_{i=1}^n | x_i|` eqn (32) in linked paper and immediately afterwards they state that MLE = MME and they also give a distributional result. 
^3 the MLE for Skellam is a special case of the root finding problem considered in section 3.2 of that paper, corresponds to COM-Poisson with the \nu_i =1 for i =1,2

Boltzmann note: Boltzmann is a truncated geometric distribution. An MLE for the probability of a geometric exists, so incorporation of a truncation point should be straightfoward. Estimation of truncation point is not known (to me) but also straightforward to work out. Joint estimation of truncation point and probability would need to be worked out. Unclear to me at this time if there are technical difficulties in this joint case. 

Note: the remaining distributions: logseries, hypergeometric, planck, zipf, may exist but I was unable to find references. 

Existence of estimators in discrete case is better than I expected but note that if a `fit` method is added this could preclude inclusion of some distributions. For example, there is an open PR with a poisson-binomial distribution, AFAIK there is no known MLE for poisson-binomial. 

After doing the research I've gone from a -0.0 on this to a +0.3 on adding the fit. I'd recommend to open a discussion thread on the topic on the  scipy-dev list. You'll reach more people there than here and get better opinions than mine. :)

**Comment by psureshmagadi17** [comment]
Hey @rlucas7 / @mdhaber / @jmmanley -  I assume there is no functionality to use ".fit" for fitting data to discrete distributions yet. Is there any other alternative that's available?

**Comment by mdhaber** [comment]
There is not. If you send data and a distribution, i might be able to throw something together that uses one of the global optimizers, though.

**Comment by mdhaber** [comment]
In the meantime:
```
import numpy as np
from scipy.optimize import brute, differential_evolution
from scipy.stats import binom
import warnings

def func(free_params, *args):
    dist, x = args
    # negative log-likelihood function
    a = -np.log(dist.pmf(x, *free_params)).sum()
    if np.isnan(a):
        a = np.inf
    return a

def fit_discrete(dist, x, bounds, optimizer=brute):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        return optimizer(func, bounds, args=(dist, x))

n, p = 5, 0.4
x = binom.rvs(n, p, size=10000)

bounds = [(0, 100), (0, 1)]
u2, s2 = fit_discrete(binom, x, bounds)
res = fit_discrete(binom, x, bounds, optimizer=differential_evolution)
print(u2, s2)
print(res.x)
```

**Comment by psureshmagadi17** [comment]
@mdhaber - unfortunately, I cannot share the data. But, I'll try to run the above code and check how it works. 

Thanks again!

**Comment by tirthasheshpatel** [comment]
@mdhaber I was experimenting with this a little and basically, the difference between using global optimization methods and analytical methods is that global optimization takes time in the order of a few seconds while analytical methods are hundreds/thousands of times faster. Also, there are a few edge cases where the global optimization method converges closer to actual parameters. Following @rlucas7's reference on estimating the `n` and `p` parameters of a binomial distribution, I tried to implement code for it [here](https://github.com/tirthasheshpatel/random_python_things/blob/master/fitting_discretedists/Fitting_Discrete_Distributions.ipynb) and got the following results:

The code for the analytical method:

```python
from scipy.stats import beta

# analytical MLE method for fitting the binomial distribution.
def fit_binom(x, alpha=0.5, bias_correction=True):
    s2 = x.var()
    x_bar = x.mean()
    xk = np.max(x)
    k = x.size
    # first moment estimate of n for a given 'alpha'
    n_hat_prior = xk**(alpha+1)*s2**alpha / (x_bar**alpha*(xk-x_bar)**alpha)
    if bias_correction:
        n_hat_prior = np.floor(n_hat_prior)
        n_hat_priors = np.arange(0, n_hat_prior-1)
        # final estimate for n after correction for bias
        n_hat = xk + np.sum(beta.ppf(1./k, n_hat_priors+1, n_hat_prior-n_hat_priors))
    else:
        n_hat = n_hat_prior
    p_hat = x_bar/n_hat

    return n_hat, p_hat
```

The estimates obtained for each of the methods with `n = 5, p = 0.4, k (sample size) = 10000` are:

```
Brute Optimization Method                  : [n = 5.00000411, p = 0.39833459]
Differential Evolution Method              : [n = 5.07169877, p = 0.39453257]
Analytical Method (with bias correction)   : [n = 5.03357740, p = 0.39566293]
Analytical Method (without bias correction): [n = 4.99194030, p = 0.39896310]
```

The time taken by each one of them is:

```
Brute Optimization Method                  : 1.77 s ± 78.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Differential Evolution Method              : 2.08 s ± 453 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Analytical Method (with bias correction)   : 857 µs ± 75.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Analytical Method (without bias correction): 299 µs ± 9.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

I tried with different combinations of `n, p, and k` and observed the following:
  - Estimate of `n` when actual `p` is small (`0.1, 0.2, 0.3`) using the analytical method: for no bias correction, it is underestimated and for bias correction, it overshoots. This is mentioned in the paper and is one of the limitations of the analytical method. Global optimizers are better here.
  - For larger `p` (`0.4, 0.5, 0.6, 0.7`), the analytical method with bias correction performs better even for small sample sizes (the estimate of `n` is between actual n +/- 10 even with sample sizes as small as 10.)
  - For values of `p` near `1`, the analytical method with no bias correction is better even for small sample sizes (`n` overshoots really badly when bias correction is performed). Global Optimizers perform as expected, poor for small sample sizes and really good for large sample sizes.

([this](https://github.com/tirthasheshpatel/random_python_things/blob/master/fitting_discretedists/fitting_discrete.csv) is the csv file of results)

If this is helpful, I can help start addressing this issue by adding a fit method for the binomial distribution.

**Comment by mdhaber** [comment]
It might be. But personally I'd want to add a generic optimizer first. I was experimenting more with that, and it would really help to have a few enhancements to global optimizers that seem like they'd be worthwhile to bake into the functions themselves rather than hacking them in just for use with discrete fit:

* Integrality constraints. When polishing is done by a local optimizer, integers in the solution would remain fixed.
* Some sort of iterative application of brute would be cool - start with a coarse grid then zoom in on the most promising region.

**Comment by casibus** [comment]
I support the idea of a generic fitting approach. 
What I did in the past for my own discret custom scipy distributions is to use a differential_evolution minimization on the negative log likelihood function as the most generic case (just the sum of the log of the pmf). That is basically @mdhaber's approach.
However, in some cases the analytical log likelihood function can be optimized quite a bit for a significant speedup. That may be achieved by applying some math and clever numpy usage. 

Take **Poisson** for example:
`logL = sum(log(mu**k / k! * exp(-mu))) = sum( k*log(mu) - log(k!)) -N*mu`

So instead of computing multiple powers and divisions using poisson.pmf() to compute the likelihood of the data we can compute it much quicker:

```
import math
import numpy as np
from scipy.special import gammaln, gamma
 
k = np.random.randint(100, size=1000)
mu = 50.

print(sum([math.log(mu**k_ / math.factorial(k_) * math.exp(-mu)) for k_ in k]))
print(sum(np.log(mu**k / gamma(k+1) * math.exp(-mu))))
print(sum(k*math.log(mu) - gammaln(k+1)) - len(k)*mu)
# --> prints three times the same value as expected

# And now let's check the timing:
%timeit sum([math.log(mu**k_ / math.factorial(k_) * math.exp(-mu)) for k_ in k])
%timeit sum(np.log(mu**k / gamma(k+1) * math.exp(-mu)))
%timeit sum(k*math.log(mu) - gammaln(k+1)) - len(k)*mu
```

As expected the last form needs least computational ressources due to lower number of operations:
```
7.2 ms ± 74.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
224 µs ± 6.11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
140 µs ± 3.03 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Obviously the MLE for Poisson is even faster as it is just computing the mean of the data ones.
In many other cases the analytical MLE exists and therefore should definetively be provided as well if possible!

But for fair comparisons of fit results of different distributions you may want to keep the fitting functionality even when an analytical solution exists.
In general, for any fitting of complex distributions the bounds of the parameters need to be specified and be checked via the `_argcheck` method.

**Comment by mdhaber** [comment]
Cool. Thanks @casibus!
The other update here is the possibility of adding biteopt to SciPy (gh-14750, gh-15298), which could work well for integer constrained variables. I planned to come back to this topic as soon as we get "official" support for MIP in one of the global optimizers.

**Comment by casibus** [comment]
To be fair, I just compared `sum(np.log(poisson.pmf(k, mu))` which I forgot earlier and it is actually not too slow:
```
7.3 ms ± 418 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) - raw python
300 µs ± 9.88 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) - sum(log(poisson.pmf())
239 µs ± 18.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each) - first numpy simplification
145 µs ± 1.71 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each) - further simplification + numpy
```

Mixed-Integer Programming sounds like a good idea!! Looking forward.

**Comment by josef-pkt** [comment]
My suggestion here would be to only support `fit` for fixed support.

Those are the most common use cases for most discrete distributions.
For example I cannot remember a usecase where standard count distributions don't start at zero, unless they are left truncated distributions. Even then, the truncation point is in almost all cases assumed to be known.

Similarly, the number of trials in binom and betabinom is usually known from the experimental design.

**Comment by mdhaber** [comment]
I just submitted gh-15389 to add integrality constraints to `brute`. Combined with the iterative refinement idea [above](https://github.com/scipy/scipy/issues/11948#issuecomment-794044761), it works really well for fitting discrete (and continuous) distributions to data. (Remember that most distributions have only a few parameters, so the curse of dimensionality doesn't hit very hard here, and with the iterative refinement idea, we never need `brute` to have a very fine grid to get accurate results, as long as the initial `bounds` are reasonable.)

What do you think about a _function_, `scipy.stats.fit(dist, data, shape_bounds, optimizer=None)` where:

- `dist` is an `rv_continuous` or `rv_discrete` distribution
- `data` is the data to be fit
- `shape_bounds` (name up for discussion) are the lower and upper bounds for each shape parameter (probably should add support for `loc` and `scale` somehow)
- `optimizer` (optional) uses the iterative brute idea by default or accepts another callable that satisfies some prescribed interface, giving the user complete control over the optimization

We'd add information about which shapes must be integral to all the distributions so the user doesn't have to.

The reason I'd rather create a new function than add a method to the `rv_discrete` class is that it would also be useful to have this function when `rv_continuous.fit` is having trouble, but it would be a bit of a headache to work this into `rv_continuous.fit`. In a SciPy 2.0, we can work this into the `fit` methods of overhauled `rv_continuous` and `rv_discrete` distributions.

**Comment by casibus** [comment]
Cool! I like the suggestion and I think it is a great idea! 
I would rather call the function something like `fit_dist` as in the context of a statistical package you could fit many other things as well (like statistical models). 
Also the proposed "fine graining" may lead to problems whenever the likelihood function is not concave (or the negative likehood convex). In other words if the likelihood landscape is like a mountain range: Whenever the global maximum is one small spiky local optimum that lies in between the grid points it will often not be found. In practice, however, it may not an issue with the standard distributions (maybe negbin and few others) but should be mentioned in the docstring. In such cases the user may just switch to another optimizer.

I would also allow to pass the `workers` and the `Ns` argument from `fit_dist `to the optimizer via `**kwds` as that can make a huge difference:

```
def fit_discrete(
        dist: Optional[scipy.stats.rv_discrete, scipy.stats.rv_continuous],
        x: Iterable[float],
        bounds: List[Tuple],
        optimizer: Callable = brute,
        **kwds,
):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            return optimizer(func, bounds, args=(dist, x), **kwds )
```

**Comment by mdhaber** [comment]
Certainly this optimizer will not work for all problems. This would just be the default because `brute` will likely be the first to support integrality constraints.

We can discuss whether to accept `kwds` and pass them to the optimizer, but I'd suggest that we just not deal with that. The user will likely need to wrap the optimizer in a very simple function (e.g. lambda) to get the format of the output to match the interface ( e.g. `brute` doesn't return an `OptimizeResult`, others do), so they can take that opportunity to pass other arguments into the optimizer, too.

**Comment by andyfaff** [comment]
See #15392 for inclusion of `integrality` constraints into `differential_evolution`.

**Comment by tirthasheshpatel** [comment]
> so I think I can draft `fit`/`fit_dist` this weekend.

Looking forward to it! Being able to fit discrete distributions will be a very useful feature! I am +1 on a function to do it. But, in long term, it will be nice to also have a `.fit` method.

**Comment by tupui** [comment]
Would be a very useful feature. Thanks for the ping, now I got the integrality PRs 😅

**Comment by mdhaber** [comment]
I checked the suitability of adding `integrality` constraints to `direct` by wrapping it to adjust bounds and round decision variables (as we did for `differential_evolution`). The results are OK, but perhaps the algorithm is not well-suited to this, because with the integrality constraints it's quite a bit slower than `differential_evolution` and literature `brute`.

<details>

Use this wrapper to add integrality constraints to a (continuous variable) global optimizer:

def _integral_wrapper(optimizer):
    def wrapped_optimizer(func, bounds, *args, integrality=None, **kwds):
        if not np.any(integrality):
            return optimizer(func, bounds, *args, **kwds)

        lb, ub = np.array(bounds, copy=True).T
        n_vars = lb.size
        integrality = np.broadcast_to(integrality, (n_vars,)).astype(bool)
        lb = np.ceil(lb)
        ub = np.floor(ub)
        if not (lb[integrality] <= ub[integrality]).all():
            message = ("No integers exist within the bounds of an integral "
                       "decision variable.")
            raise ValueError(message)
        nlb = np.nextafter(lb[integrality] - 0.5, np.inf)
        nub = np.nextafter(ub[integrality] + 0.5, -np.inf)
        lb[integrality] = nlb
        ub[integrality] = nub
        bounds = np.asarray([lb, ub]).T.tolist()

        def wrapped_func(x, *args):
            x = x.copy()
            x[integrality] = np.round(x[integrality])
            return func(x, *args)

        res = optimizer(wrapped_func, bounds, *args, **kwds)
        res.x[integrality] = np.round(res.x[integrality])
        return res
    return wrapped_optimizer

Here are the results of the test above with `optimizer = _integral_wrapper(direct)`.

![image](https://user-images.githubusercontent.com/6570539/149646494-73f52c79-47e1-4859-a262-5380c82ef117.png)

</details>

**Comment by dschmitz89** [comment]
> I checked the suitability of adding `integrality` constraints to `direct` ([gh-14300](https://github.com/scipy/scipy/pull/14300)) by wrapping it to adjust bounds and round decision variables (as we did for `differential_evolution`). The results are OK, but perhaps the algorithm is not well-suited to this, because with the integrality constraints it's quite a bit slower than `differential_evolution` and literature `brute`.
> 
> Use this wrapper to add integrality constraints to a (continuous variable) global optimizer:
> 
> def _integral_wrapper(optimizer): def wrapped_optimizer(func, bounds, *args, integrality=None, **kwds): if not np.any(integrality): return optimizer(func, bounds, *args, **kwds)
> 
> ```
>     lb, ub = np.array(bounds, copy=True).T
>     n_vars = lb.size
>     integrality = np.broadcast_to(integrality, (n_vars,)).astype(bool)
>     lb = np.ceil(lb)
>     ub = np.floor(ub)
>     if not (lb[integrality] <= ub[integrality]).all():
>         message = ("No integers exist within the bounds of an integral "
>                    "decision variable.")
>         raise ValueError(message)
>     nlb = np.nextafter(lb[integrality] - 0.5, np.inf)
>     nub = np.nextafter(ub[integrality] + 0.5, -np.inf)
>     lb[integrality] = nlb
>     ub[integrality] = nub
>     bounds = np.asarray([lb, ub]).T.tolist()
> 
>     def wrapped_func(x, *args):
>         x = x.copy()
>         x[integrality] = np.round(x[integrality])
>         return func(x, *args)
> 
>     res = optimizer(wrapped_func, bounds, *args, **kwds)
>     res.x[integrality] = np.round(res.x[integrality])
>     return res
> return wrapped_optimizer
> ```
> 
> Here are the results of the test above with `optimizer = _integral_wrapper(direct)`.
> 
> ![image](https://user-images.githubusercontent.com/6570539/149646494-73f52c79-47e1-4859-a262-5380c82ef117.png)

Nice. I guess DIRECT takes longer because of the very tight default tolerances. Could be incoroporated into DIRECT in a follow-up PR.

**Comment by antelk** [comment]
> In the meantime:
> 
> ```
> import numpy as np
> from scipy.optimize import brute, differential_evolution
> from scipy.stats import binom
> import warnings
> 
> def func(free_params, *args):
>     dist, x = args
>     # negative log-likelihood function
>     ll = -np.log(dist.pmf(x, *free_params)).sum()
>     if np.isnan(ll):  # occurs when x is outside of support
>         ll = np.inf   # we don't want that
>     return ll
> 
> def fit_discrete(dist, x, bounds, optimizer=brute):
>     with warnings.catch_warnings(): 
>         warnings.simplefilter("ignore")  
>         return optimizer(func, bounds, args=(dist, x))
> 
> n, p = 5, 0.4
> x = binom.rvs(n, p, size=10000)
> 
> bounds = [(0, 100), (0, 1)]
> u2, s2 = fit_discrete(binom, x, bounds)
> res = fit_discrete(binom, x, bounds, optimizer=differential_evolution)
> print(u2, s2)
> print(res.x)
> ```
> 
> ```
> 5.000011549883961 0.3978198921753501
> [5.01158844 0.39695459]
> ```
> 
> I am curious how well something so simple works, so if you try it out, let me know how it goes.

Hi everybody,

I've been playing around with the fitting of discrete distributions in my code. Since I am required to use `scipy`'s version <1.8.x because another dependency, I am using the code similar to the one quoted before to fit a set of discrete distributions.
I tried to use this code to fit discrete distribution with 1.9.1 version of `scipy` and the fitted parameters do not converge.
I also tried to use this same code to fit discrete distribution with the most recent version of `scipy`, and the same thing happens:
```ipython
In [...]: print(u2, s2); print(res.x)
105.0 0.01906476201195455
[12.28294017  0.5646521 ]
```

Since I am really not well versed in `scipy.optimize` module, does anybody have any idea why is this the case? Why does the new version of `scipy` won't converge and the one before `scipy.stats.fit` was added converges to pretty accurate values?

Thank you!

**Comment by mdhaber** [comment]
Can you share the code that reproduces the issue?

**Comment by antelk** [comment]
Unfortunately, I cannot share the code I am using.

I can maybe cook another example, but the code snippet at the beginning of my question above gives these strange values by using the version 1.7.3 of `scipy`.

`scipy.stats.fit` works great.

**Comment by mdhaber** [comment]
`binom.pmf` changed and now seems to return `inf` when evaluated at non-integral `n`. 
I added  `free_params[0] = np.round(free_params[0])` into `func` and it works (but of course returns non-integral `n`, which you'll need to round).
I'll fix `binom`.

**Comment by antelk** [comment]
Thank you for the clarification!
It really does work now.
